### PR TITLE
[feature] Added a context id to called functions

### PR DESCRIPTION
**Pull request by naranjamecanica**
We use multipage setup a lot in our game projects. To be able to provide each game page with its own data i've added a pageId parameter to the called context functions. See README.md and tests for the details.

Hope you can merge this! 👍🏻

### CODE REVIEW

File: src/context.ts [hunk 1]
```diff
@@ -20,11 +24,29 @@ export async function resolveContext(
...
+      output[key] = value;
     }
   }
 
   return output;
 }
+
+export function determineContextPageId(ctx: IndexHtmlTransformContext): string {
+  // make sure paths are consistent on win32/posix
+  const fullFilePath = ctx.path.split(path.sep).join(path.posix.sep);
```
Comment (alexlafroscia) [inline src/context.ts:38]: I've seen this `normalizePath` helper in the Vite docs -- would it make sense to use that here?

https://vitejs.dev/guide/api-plugin.html#path-normalization

File: src/index.ts [hunk 2]
```diff
@@ -1,8 +1,9 @@
 import { compile, registerHelper, RuntimeOptions } from 'handlebars';
 import { resolve } from 'path';
 import { Plugin as VitePlugin } from 'vite';
+import type { IndexHtmlTransformContext } from 'vite';
```
Comment (alexlafroscia) [inline src/index.ts:4]: I think this lines can be combined like so:

```suggestion
import { Plugin as VitePlugin, IndexHtmlTransformContext } from 'vite';
```

**Review by alexlafroscia — COMMENTED** [review]
Thanks for the PR! I'm overall happy to accept this feature, as I can totally see it being useful.

I'm curious if the page ID should include the file extension, though. In other places within Vite there a page ID is used to reference a file, I believe the extension is left in tact. Maybe it makes things bit simpler if we just pass `ctx.path` along "untouched" instead?

Let me know what you think about that!

File: README.md [hunk 3]
```diff
@@ -64,7 +64,31 @@ export default {
 
 This will result in `<h1>Hello, world!</h1>` in your output HTML file.
 
-You can also provide a function, either as the `context` key or any of the keys within the object, which will be evaluated to create the value that will be made available inside your application.
+You can also provide a (asynchronous) function, either as the `context` key or any of the keys within the object, which will be evaluated to create the value that will be made available inside your page. This function is called with an identifier parameter based on the HTML file path which makes it possible to provide unique data to each HTML page in a multipage application setup.
```
Comment (alexlafroscia) [inline README.md:67]:
```suggestion
You can also provide an (asynchronous) function, either as the `context` key or any of the keys within the object, which will be evaluated to create the value that will be made available inside your page. This function is called with an identifier parameter based on the HTML file path which makes it possible to provide unique data to each HTML page in a multipage application setup.
```

File: src/context.ts (hunk 1 above)
Comment (naranjamecanica) [reply src/context.ts:38]: I've tried to implement this, but microbundle chokes on some part in vite. 

```
microbundle -f es,cjs --target node --no-compress

Error: Unexpected token

at /Users/naranjamecanica/Developer/opensource/vite-plugin-handlebars/node_modules/vite/dist/node/chunks/dep-0776dd57.js:68846:60

68844: 
68845: const pendingModules = new Map();
68846: async function ssrLoadModule(url, server, context = { global }, urlStack = []) {                                            
```

File: src/index.ts (hunk 2 above)
Comment (naranjamecanica) [reply src/index.ts:4]: I will change this, but using `import type` makes sure no unused imports are left over in the build js file.

**Comment by naranjamecanica** [comment]
I've updated my code to use a path, but was not able to implement the vite method.  See: https://github.com/alexlafroscia/vite-plugin-handlebars/pull/42#discussion_r601820517

**Comment by alexlafroscia** [comment]
Hmm, it seems like CI didn't run for this... I'm not sure why...

**Comment by alexlafroscia** [comment]
@naranjamecanica can you rebase from `main`? I updated our CI action config to run on `pull_request` events as well as `push`; I'm not sure why the `push` for your commit earlier didn't trigger the tests

File: README.md [hunk 4]
```diff
@@ -64,7 +64,31 @@ export default {
...
+
+```javascript
+// vite.config.js
+import handlebars from 'vite-plugin-handlebars';
+
+const pageData = {
+  index: {
```
Comment (alexlafroscia) [inline README.md:74]: These should be updated to include the file extension, since that's what's exposed to the function

File: README.md [hunk 5]
```diff
@@ -64,7 +64,31 @@ export default {
...
+  },
+  subpage: {
+    title: 'Sub Page',
+  }
+}
+
+export default {
+  plugins: [
+    handlebars({
+      context: (pagePath) {
```
Comment (alexlafroscia) [inline README.md:85]: This syntax isn't valid -- either dropping the `:` or making the function an arrow-function should fix it!

```suggestion
      context(pagePath) {
```

File: __tests__/context.ts [hunk 6]
```diff
@@ -60,3 +60,20 @@ test('it evaluates an asynchronous `context` function', async () => {
 
   expect(html).toContain('<p>bar</p>');
 });
+
+test('it evaluates an asynchronous `context` function using an id', async () => {
+  const temp = await factory.createStructure({
+    'index.html': '<p>{{foo}}</p>',
+  });
+  const data: Record<string, Record<string, unknown>> = {
+    index: {
```
Comment (alexlafroscia) [inline __tests__/context.ts:69]: This needs to include the file extension as well; the test is failing locally

**Review by alexlafroscia — CHANGES_REQUESTED** [review]

**Comment by alexlafroscia** [comment]
Thanks for the quick follow-up there!

This has been released as `v1.5.0` 🎉

**Pull request merged**